cctp_primitives/src/utils/proving_system.rs

```diff
@@ -81,15 +286,27 @@ pub fn generate_circuit_keypair<C: ConstraintSynthesizer<FieldElement>>(
         ProvingSystem::Undefined => return Err(ProvingSystemError::UndefinedProvingSystem)?,
         ProvingSystem::CoboundaryMarlin => {
             let (pk, vk) = CoboundaryMarlin::setup(circ)?;
-            write_to_file(&pk, pk_path)?;
-            write_to_file(&vk, vk_path)?;
+            write_to_file(&ZendooProverKey::CoboundaryMarlin(pk), pk_path)?;
+            write_to_file(&ZendooVerifierKey::CoboundaryMarlin(vk), vk_path)?;
         },
         ProvingSystem::Darlin => {
             let (pk, vk) = Darlin::setup(circ)?;
-            write_to_file(&pk, pk_path)?;
-            write_to_file(&vk, vk_path)?;
+            write_to_file(&ZendooProverKey::Darlin(pk), pk_path)?;
+            write_to_file(&ZendooVerifierKey::Darlin(vk), vk_path)?;
         },
     }
 
     Ok(())
+}
+
+/// Utility function: check that proof and vk belong to the same proving system.
+pub fn check_matching_proving_system_type(
+    proof: &ZendooProof,
+    vk:    &ZendooVerifierKey,
```

Comment:
During SC declaration in MC node we currently have 2 specific arguments: "provingSystem" and "vk". With the new approach we should change MC logic and pass only 1 parameter (with both proving system type and vk), right?
How then MC node will know what "proof_verify" method to call: Darlin or CobMarlin? Probably, we need to add the "get_proving_system_type" method for Vk at least.

It seems, that previous approach may be even easier to maintain.
@alsala what do you think about the new approach?

Reply:
Proofs and vks will be serialized/deserialized with an additional byte indicating the proving system.
MC will call a single verify_proof function (for the batch verifier it's not necessary since he accepts both Darlin and CobMarlin proof), and this function, internally, will take care of choosing the appropriate verifier given the type of the vk and the proof (that must be the same of course). Look at the logic to create/verify the threshold sig proof in zendoo-sc-cryptolib (that I've already updated with these changes) to look at how it's done

Reply:
This is beneficial mostly for the mc-cryptolib, allowing to not duplicating functions or passing the ProvingSystem type to every function

Reply:
Maybe we can add to the is_valid() functions for proofs and vks a check enforcing that the type of the proof/vk is equal to the one declared in the scc

Reply:
> 
> 
> During SC declaration in MC node we currently have 2 specific arguments: "provingSystem" and "vk". With the new approach we should change MC logic and pass only 1 parameter (with both proving system type and vk), right?
> How then MC node will know what "proof_verify" method to call: Darlin or CobMarlin? Probably, we need to add the "get_proving_system_type" method for Vk at least.
> 
> It seems, that previous approach may be even easier to maintain.
> @alsala what do you think about the new approach?

We could probably keep both arguments, even if they could be redundant, and using the vk.IsValid(type) as an extra check.